text/0000-exit-codes.md

```diff
@@ -0,0 +1,117 @@
+- Feature Name: exit_codes
+- Start Date: 2018-06-11
+- RFC PR: (leave this empty)
+- Notion Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Exit codes are an important part of Notion's communication with external tools and a first-class part of the design of its user-friendly error messages.  We need a strategy for ensuring they are used and communicated consistently.
+
+# Motivation
+[motivation]: #motivation
+
+Each exit code is intended to signal to an external process some information about Notion's state at the end of its execution.  Ideally, we might wish to have a separate exit code for each error Notion is capable of producing, but many modern Unixes (including macOS) only support eight bits of exit status.  If the number of unique exit conditions in Notion exceeds that, we'll need to start "doubling up".
+
+Because redefining exit codes later in Notion's life would be a major hassle for third-party tools reliant on them, we should consider at the outset grouping errors together by exit code.
+
+Further, Notion must avoid conflicts between the exit codes produced by shims and those produced by the shimmed executables, in order to best preserve their behavior.
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+A list of exit codes must be maintained within the Notion source code for use by developers, which should contain descriptions of when each code should be used.  This same list can be exposed via consumer-facing documentation to ensure that Notion developers and consumers share an understanding of the meaning of each exit code.
+
+# Details
+[details]: #details
+
+## Shims
+
+Of the three "first-class" shims (`node`, `npm`, and `yarn`), only Node currently publishes a [list of exit codes](https://nodejs.org/api/process.html#process_exit_codes); the other two only emit codes 0/1 for success/failure.  The exit codes for other shimmed executables is too large a space to examine, but Node follows what appears to be a common practice: simply numbering errors starting with 1, preserving the [special meanings commonly applied by POSIX shells](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_08_02).  This gives us incentive to avoid low-numbered exit codes (as likely to conflict) and exit codes with the high bit set (as typically representing termination due to signals).
```

Comment:
Don't forget `npx`, too.

Reply:
Is `npx` a first-class shim (i.e. does it have special handling attached to it)?  I had thought it was being handled as a "generic" 3rd-party shim.